adaptris/interlok-aws
# INTERLOK-3465 S3 RetryStore Impl

## Motivation

This is the AWS-S3 implementation for https://github.com/adaptris/interlok/blob/develop/docs/adr/0008-restful-failed-message-retrier.md

The intention is to have a pluggable s3-retry-store implementation that duplicates the reference filesystem behaviour made available as part of https://github.com/adaptris/interlok/pull/556. 

## Modification

- Added a S3RetryStore implementation
- Refactor the localstack tests to expose shared functionality
- Mockito based tests for 100% coverage.
- Change the existing behaviour in interlok-json (https://github.com/adaptris/interlok-json/commit/43e0bb11a3251c0082d71ea5f964088aeb50d60e) so that bucket is not emitted if it's null.
- Change the existing behaviour in interlok-core (https://github.com/adaptris/interlok/commit/0cb7262102273c46acc3c070a5340b8b90fd05b3) so that we don't emit the bucket.
- The layout duplicates that of the filesystem-retry-store, so separate metadata.properties / payload.blob; and handles the AWS nuance around _trying to think about prefixes as directories even though they aren't_

The rationale behind not emitting the bucket is to remove any ambiguity around what's returned by the `report()` operation. We don't expose anything information other than _what could be retried via the jetty-retrier_... rather than letting people know about buckets / directories (since we drop the prefix when doing the S3 report).

## Result

- `amazon-s3-retry-store` is now selectable as a retry store implementation; (requires 3.11-SNAPSHOT).
- RemoteBlobIterable is now a public class not package level (though it possibly doesn't need to be).

## Testing

Based on the configuration detailed in : https://github.com/adaptris/interlok/pull/556 we should be able to do this

- Add amazon-s3 as a shared connection
```xml
      <amazon-s3-connection>
        <unique-id>shared-aws-s3</unique-id>
        <region>${amazon.region}</region>
        <credentials class="aws-static-credentials-builder">
          <authentication class="aws-keys-authentication">
            <access-key>${amazon.access.key}</access-key>
            <secret-key>${amazon.secret.key}</secret-key>
          </authentication>
        </credentials>
      </amazon-s3-connection>
```
- Change the retry store implementation in the `retry-store-write-message` and `retry-via-jetty` implementations.
```xml
          <retry-store class="amazon-s3-retry-store">
            <connection class="shared-connection">
              <lookup-name>shared-aws-s3</lookup-name>
            </connection>
            <bucket>${amazon.s3.bucket}</bucket>
            <prefix>${amazon.s3.prefix}</prefix>
          </retry-store>
```

- This now gives you (if you make a message fail first of all)

```console
$ curl -XGET -si http://localhost:8080/api/failed/list
HTTP/1.1 200 OK
Content-Type: application/json
Transfer-Encoding: chunked

[{"size":11,"name":"060dbc35-affe-4a10-b504-520edd0e2843","lastModified":1604400580000}]
chanl3@lhrm32445 ~/work/dev/aws/interlok-aws-s3 (INTERLOK-3465-S3-RetryStore-Impl)
$ curl -XGET -si http://localhost:8080/api/failed/list

chanl3@lhrm32445 ~/work/dev/aws/interlok-aws-s3 (INTERLOK-3465-S3-RetryStore-Impl)
$ curl -XPOST -si http://localhost:8080/api/retry/060dbc35-affe-4a10-b504-520edd0e2843
HTTP/1.1 202 Accepted
Content-Type: text/plain
Transfer-Encoding: chunked


```
